src/core/ext/transport/chttp2/transport/writing.cc

```diff
@@ -174,343 +174,456 @@ static bool is_default_initial_metadata(grpc_metadata_batch *initial_metadata) {
-      t->settings[GRPC_PEER_SETTINGS][GRPC_CHTTP2_SETTINGS_HEADER_TABLE_SIZE]);
+  void EnactHpackSettings(grpc_exec_ctx *exec_ctx) {
+    grpc_chttp2_hpack_compressor_set_max_table_size(
+        &t_->hpack_compressor,
+        t_->settings[GRPC_PEER_SETTINGS]
+                    [GRPC_CHTTP2_SETTINGS_HEADER_TABLE_SIZE]);
+  }
 
-  if (t->flow_control.remote_window > 0) {
-    while (grpc_chttp2_list_pop_stalled_by_transport(t, &s)) {
-      if (!t->closed && grpc_chttp2_list_add_writable_stream(t, s)) {
-        stream_ref_if_not_destroyed(&s->refcount->refs);
+  void UpdateStreamsNoLongerStalled() {
+    grpc_chttp2_stream *s;
+    while (grpc_chttp2_list_pop_stalled_by_transport(t_, &s)) {
+      if (!t_->closed && grpc_chttp2_list_add_writable_stream(t_, s)) {
+        if (!stream_ref_if_not_destroyed(&s->refcount->refs)) {
+          grpc_chttp2_list_remove_writable_stream(t_, s);
+        }
       }
     }
   }
 
-  grpc_chttp2_begin_write_result result = {false, false, false};
+  grpc_chttp2_stream *NextStream() {
+    if (t_->outbuf.length > target_write_size(t_)) {
+      result_.partial = true;
+      return nullptr;
+    }
 
-  /* for each grpc_chttp2_stream that's become writable, frame it's data
-     (according to available window sizes) and add to the output buffer */
-  while (true) {
-    if (t->outbuf.length > target_write_size(t)) {
-      result.partial = true;
-      break;
+    grpc_chttp2_stream *s;
+    if (!grpc_chttp2_list_pop_writable_stream(t_, &s)) {
+      return nullptr;
+    }
+
+    return s;
+  }
+
+  void ResetPingRecvClock() {
+    if (!t_->is_client) {
+      t_->ping_recv_state.last_ping_recv_time = GRPC_MILLIS_INF_PAST;
+      t_->ping_recv_state.ping_strikes = 0;
+    }
+  }
+
+  void IncInitialMetadataWrites() { ++initial_metadata_writes_; }
+  void IncWindowUpdateWrites() { ++flow_control_writes_; }
+  void IncMessageWrites() { ++message_writes_; }
+  void IncTrailingMetadataWrites() { ++trailing_metadata_writes_; }
+
+  void NoteScheduledResults() { result_.early_results_scheduled = true; }
+
+  grpc_chttp2_transport *transport() const { return t_; }
+
+  grpc_chttp2_begin_write_result Result() {
+    result_.writing = t_->outbuf.count > 0;
+    return result_;
+  }
+
+ private:
+  grpc_chttp2_transport *const t_;
+
+  /* stats histogram counters: we increment these throughout this function,
+     and at the end publish to the central stats histograms */
+  int flow_control_writes_ = 0;
+  int initial_metadata_writes_ = 0;
+  int trailing_metadata_writes_ = 0;
+  int message_writes_ = 0;
+  grpc_chttp2_begin_write_result result_ = {false, false, false};
+};
+
+class DataSendContext {
+ public:
+  DataSendContext(WriteContext *write_context, grpc_chttp2_transport *t,
+                  grpc_chttp2_stream *s)
+      : write_context_(write_context),
+        t_(t),
+        s_(s),
+        sending_bytes_before_(s_->sending_bytes) {}
+
+  uint32_t stream_remote_window() const {
```

Comment:
Tiny nit; why are stream_remote_window and max_outgoing snake case, but the rest of the new C++ camel?

Reply:
https://google.github.io/styleguide/cppguide.html#Function_Names

accessors and setters get snake_case, methods get CamelCase (see the final paragraph).